sql/2021/pwa/manifests_not_json_parsable.sql

```diff
@@ -0,0 +1,39 @@
+#standardSQL
+# Manifests that are not JSON parsable for all pages
```

Comment:
What is the goal of this report? Is it to measure the number of "invalid" Manifests, or is there any other reason for it?

Reply:
Not sure to be honest. It was copied from last two years queries.

But yeah good to see malformed manifests are mostly valid JSON.

Might also be worth looking at 404 ones? I.e. where a manifest is referenced in the HTML but not actually served (or an HTML 404 page served instead of a JSON 200 ).

Reply:
Had a look at the 404 suggestion and 99.86% of sites return a 200. So not that interesting, so won't bother including. Will leave you to decide whether you wanna use the existing stat or not. Resolving.

client | status | freq | total | pct
-- | -- | -- | -- | --
desktop |   | 122 | 98,655 | 0.12%
desktop | 304 | 45 | 98,655 | 0.05%
desktop | 200 | 98,488 | 98,655 | 99.83%
mobile | 404 | 1 | 128,541 | 0.00%
mobile | - | 5 | 128,541 | 0.00%
mobile | 200 | 128,364 | 128,541 | 99.86%
mobile |   | 118 | 128,541 | 0.09%
mobile | 304 | 53 | 128,541 | 0.04%

```sql
CREATE TEMP FUNCTION jsonObjectKeys(input STRING)
RETURNS Array<String>
LANGUAGE js AS """
  try {
    return Object.keys(JSON.parse(input));
  } catch {
    return "['Invalid']";
  }
""";

SELECT
  _TABLE_SUFFIX as client,
  status,
  COUNT(DISTINCT url) AS freq,
  SUM(COUNT(DISTINCT url)) OVER (PARTITION BY _TABLE_SUFFIX) AS total,
  COUNT(DISTINCT url) / SUM(COUNT(DISTINCT url)) OVER (PARTITION BY _TABLE_SUFFIX) AS pct
FROM
  `httparchive.pages.2021_07_01_*`,
  --`httparchive.sample_data.pages_*`,
  UNNEST(jsonObjectKeys(JSON_EXTRACT(payload, '$._pwa.manifests'))) AS manifest_url
LEFT OUTER JOIN
  (
    SELECT
      _TABLE_SUFFIX as _TABLE_SUFFIX,
      p.url,
      r.url as manifest_url,
      status
    FROM
      --`httparchive.sample_data.summary_requests_*` r
      `httparchive.summary_requests.2021_07_01_*` r
    JOIN
      --`httparchive.sample_data.summary_pages_*` p
      `httparchive.summary_pages.2021_07_01_*` p
    USING (_TABLE_SUFFIX, pageid)
  )
USING (_TABLE_SUFFIX, url, manifest_url)
WHERE
  JSON_EXTRACT(payload, '$._pwa') != "[]" AND
  JSON_EXTRACT(payload, '$._pwa.manifests') != "[]" AND
  JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = "true"
GROUP BY
  client,
  status

```